File: src/shared/types/translation.ts
export type TranslationLanguageCode = 'auto' | 'zh-CHS' | 'en' | 'ja' | 'ko' | 'fr' | 'de' | 'es' | 'ru'

export type TranslationProvider = 'youdao' | 'openai_compatible'

export type BackendAuthState = 'anonymous' | 'authenticated'

export type TranslationProviderBindingMode = 'managed' | 'byo_key'

export type TranslationProviderConfigSummary = {
  credentialHint?: string
  endpointUrl?: string
  model?: string
}

export type YoudaoProviderConfigInput = {
  provider: 'youdao'
  appKey: string
  appSecret: string
}

export type OpenAICompatibleProviderConfigInput = {
  provider: 'openai_compatible'
  baseUrl: string
  apiKey: string
  model: string
}

export type TranslationProviderConfigInput = YoudaoProviderConfigInput | OpenAICompatibleProviderConfigInput

export type TranslationPreferences = {
  defaultProvider: TranslationProvider
  sourceLanguage: TranslationLanguageCode
  targetLanguage: Exclude<TranslationLanguageCode, 'auto'>
  autoTranslateEnabled: boolean
}

export type BackendConfig = {
  baseUrl: string
  authState: BackendAuthState
  accessToken: string
  refreshToken: string
}

export type TranslationProviderStatus = {
  provider: TranslationProvider
  platformAvailable: boolean
  userConfigured: boolean
  configMode: TranslationProviderBindingMode | null
  status: 'available' | 'unavailable' | 'not_configured'
  lastErrorCode?: string
  configSummary?: TranslationProviderConfigSummary | null
}

export type TranslationResult = {
  query: string
  translation: string
  detectedSourceLanguage: string
  targetLanguage: string
  provider: TranslationProvider
}

export const PRODUCTION_BACKEND_BASE_URL = 'https://www.mingcai-colorful.top'

const resolveDefaultBackendBaseUrl = () => {
  const configuredBaseUrl = import.meta.env.VITE_BACKEND_BASE_URL?.trim()

  if (!configuredBaseUrl) {
    return PRODUCTION_BACKEND_BASE_URL
  }

  return configuredBaseUrl.replace(/\/$/, '')
}

export const DEFAULT_BACKEND_BASE_URL = resolveDefaultBackendBaseUrl()

export const DEFAULT_TRANSLATION_PREFERENCES: TranslationPreferences = {
  defaultProvider: 'youdao',
  sourceLanguage: 'auto',
  targetLanguage: 'zh-CHS',
  autoTranslateEnabled: false
}

export const DEFAULT_BACKEND_CONFIG: BackendConfig = {
  baseUrl: DEFAULT_BACKEND_BASE_URL,
  authState: 'anonymous',
  accessToken: '',
  refreshToken: ''
}

File: backend/app/schemas/translation.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import Field

from app.schemas.common import CamelModel

TranslationProvider = Literal["youdao", "openai_compatible"]
TranslationLanguageCode = Literal["auto", "zh-CHS", "en", "ja", "ko", "fr", "de", "es", "ru"]


class ClientContext(CamelModel):
    extension_version: Optional[str] = None
    page_url: Optional[str] = None
    page_title: Optional[str] = None


class TranslationRequestIn(CamelModel):
    text: str = Field(min_length=1, max_length=5000)
    source_language: TranslationLanguageCode = "auto"
    target_language: TranslationLanguageCode = "zh-CHS"
    provider_hint: Optional[TranslationProvider] = None
    client_context: Optional[ClientContext] = None


class TranslationQuota(CamelModel):
    mode: Literal["platform", "user"]
    requests_per_minute: Optional[int] = None
    daily_limit: Optional[int] = None


class TranslationResultOut(CamelModel):
    translated_text: str
    detected_source_language: str
    target_language: str
    provider: TranslationProvider
    request_id: str
    quota: TranslationQuota


class TranslationPreferenceOut(CamelModel):
    default_provider: TranslationProvider
    source_language: TranslationLanguageCode
    target_language: TranslationLanguageCode
    auto_translate_enabled: bool
    updated_at: str


class TranslationPreferenceUpdate(CamelModel):
    default_provider: TranslationProvider = "youdao"
    source_language: TranslationLanguageCode = "auto"
    target_language: TranslationLanguageCode = "zh-CHS"
    auto_translate_enabled: bool = False


class ProviderStatusOut(CamelModel):
    provider: TranslationProvider
    platform_available: bool
    user_configured: bool
    config_mode: Optional[Literal["managed", "byo_key"]] = None
    status: Literal["available", "unavailable", "not_configured"]
    last_error_code: Optional[str] = None
    config_summary: Optional["ProviderConfigSummaryOut"] = None


class ProviderConfigSummaryOut(CamelModel):
    credential_hint: Optional[str] = None
    endpoint_url: Optional[str] = None
    model: Optional[str] = None


class ProviderConfigUpdate(CamelModel):
    youdao_app_key: Optional[str] = Field(default=None, max_length=255)
    youdao_app_secret: Optional[str] = Field(default=None, max_length=255)
    openai_base_url: Optional[str] = Field(default=None, max_length=512)
    openai_api_key: Optional[str] = Field(default=None, max_length=512)
    openai_model: Optional[str] = Field(default=None, max_length=255)
